src/goto-instrument/code_contracts.cpp

```diff
@@ -282,24 +335,388 @@ const symbolt &code_contractst::new_tmp_symbol(
+      disjuncts.push_back(running);
+      running = disjunction(disjuncts);
+    }
+  }
+
+  return running;
+}
+
+void code_contractst::populate_assigns_references(
+  const symbolt &function_symbol,
+  const irep_idt &function_id,
+  goto_programt &created_decls,
+  std::vector<exprt> &created_references)
+{
+  const code_typet &type = to_code_type(function_symbol.type);
+  exprt assigns = static_cast<const exprt &>(type.find(ID_C_spec_assigns));
+
+  exprt::operandst &targets = assigns.operands();
+  for(exprt curr_op : targets)
+  {
+    exprt op_addr = address_of_exprt{curr_op};
+
+    // Declare a new symbol to stand in for the reference
+    symbol_exprt standin = new_tmp_symbol(
+                             pointer_type(curr_op.type()),
+                             function_symbol.location,
+                             function_id,
+                             function_symbol.mode)
+                             .symbol_expr();
+
+    created_decls.add(
+      goto_programt::make_decl(standin, function_symbol.location));
+
+    created_decls.add(goto_programt::make_assignment(
+      code_assignt(standin, std::move(op_addr)), function_symbol.location));
+
+    // Add a map entry from the original operand to the new symbol
+    created_references.push_back(standin);
+  }
+}
+
+void code_contractst::instrument_assn_statement(
+  goto_programt::instructionst::iterator &instruction_iterator,
+  goto_programt &program,
+  exprt &assigns,
+  std::vector<exprt> &assigns_references,
+  std::set<exprt> &freely_assignable_exprs)
+{
+  INVARIANT(
+    instruction_iterator->is_assign(),
+    "The first argument of instrument_assn_statement should always be"
+    " an assignment");
+  const exprt &lhs = instruction_iterator->get_assign().lhs();
+  if(freely_assignable_exprs.find(lhs) != freely_assignable_exprs.end())
+  {
+    return;
+  }
+  exprt alias_expr = create_alias_expression(assigns, lhs, assigns_references);
+
+  goto_programt alias_assertion;
+  alias_assertion.add(goto_programt::make_assertion(
+    alias_expr, instruction_iterator->source_location));
+  program.insert_before_swap(instruction_iterator, alias_assertion);
+  ++instruction_iterator;
+}
+
+void code_contractst::instrument_call_statement(
+  goto_programt::instructionst::iterator &instruction_iterator,
+  goto_programt &program,
+  exprt &assigns,
+  std::vector<exprt> &aliasable_references,
+  std::set<exprt> &freely_assignable_exprs)
+{
+  INVARIANT(
+    instruction_iterator->is_function_call(),
+    "The first argument of instrument_call_statement should always be "
+    "a function call");
+  code_function_callt call = instruction_iterator->get_function_call();
+  const irep_idt &called_name =
+    to_symbol_expr(call.function()).get_identifier();
+
+  // Malloc allocates memory which is not part of the caller's memory
+  // frame, so we want to capture the newly-allocated memory and
+  // treat it as assignable.
+  if(called_name == "malloc")
```

Comment:
I'd like to see handling of all uses of `__CPROVER_allocate` here. `malloc` is just one such case invoking `__CPROVER_allocate`. I'd propose a simple analysis that
1. Checks whether the body of the called function is available,
2. and if it is, whether the body of that function includes a `__CPROVER_allocate` call.

Reply:
Tracking this request in https://github.com/diffblue/cbmc/issues/5529.